### PR TITLE
test reduce and allreduce fail with custom datatype and custom operator

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**
Custom reduction operation does not seem to be working on custom datatypes. Test case added for this.

### CODE REVIEW

File: test/mpi/coll/testlist.in [hunk 1]
```diff
@@ -56,6 +56,8 @@ coll10 4
 coll11 4
 coll12 4
 coll13 4
+customreduce 1
```
Comment (akhillanger) [inline test/mpi/coll/testlist.in]: I think you would want to test it for at least 2 ranks?

Comment (KavithaTipturMadhu) [reply test/mpi/coll/testlist.in]: The tests fail for one. I will add two ranks, but the issue isn't with ranks.

Comment (akhillanger) [reply test/mpi/coll/testlist.in]: Does it fail only when number of ranks is 1? With 1 rank, reduce and allreduce should return immediately without having to do any computation/communication.

Comment (KavithaTipturMadhu) [reply test/mpi/coll/testlist.in]: The tests fail with ranks>=1.

File: test/mpi/coll/customreduce.c [hunk 2]
```diff
@@ -0,0 +1,102 @@
+#include "mpi.h"
```
Comment (raffenet) [inline test/mpi/coll/customreduce.c]: There is a copyright header that needs to be included with all files in the repo. Can you make sure to add it here?

File: test/mpi/coll/customreduce.c [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+{
+    int rank, count, size, minsize = 1;
+    int isLeftGroup;
+    MPI_Status status;
+    unsigned int errs = 0;
+    MPI_Comm comm;
+    MTestDatatype sendtype, recvtype;
+    MTest_Init(&argc, &argv);
+
+    //Reduce with custom operator
```
Comment (raffenet) [inline test/mpi/coll/customreduce.c]: We try to avoid C++ style comments if we can. Please use `/* */ for consistency.

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

File: test/mpi/coll/testlist.in (hunk 1 above)
Comment (raffenet) [reply test/mpi/coll/testlist.in]: Any more information about where the failure occurs in MPICH for this test? Some additional information would be helpful, either in the commit message or in the test comments.

File: test/mpi/coll/Makefile.am [hunk 4]
```diff
@@ -69,6 +69,7 @@ noinst_PROGRAMS =      \
     icreduce           \
     icscatter          \
     icscatterv         \
+    customreduce       \
```
Comment (raffenet) [inline test/mpi/coll/Makefile.am]: Make sure to also add the test to `test/mpi/.gitignore`.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/tcp

**Comment by akhillanger** [comment]
There is a problem with this test. When you do MTestCheckRecv(0, &recvtype); the function is going to check for data of type recvType. However, Reduce call takes only sendtype as argument and it would not have converted the reduced buffer to recvType datatype.

**Comment by akhillanger** [comment]
The correct test would be to initialize two copies of sendtype at the root - first is used in the send buffer and the second is used in the recv buffer of reduce collective. And then call MTestCheckRecv on the second copy.

**Comment by KavithaTipturMadhu** [comment]
Is the recommendation that I use the same datatype for both send and receive buffers? The tests work for same send and receive buffer types but fail when the send and receive type are different.

**Comment by akhillanger** [comment]
MPI_Reduce and Allreduce take only one datatype as an argument, which means that for root send and recv type are the same. Different ranks can have different datatypes.

**Comment by KavithaTipturMadhu** [comment]
I've fixed the test to limit to matching buffer types.

**Comment by akhillanger** [comment]
Some tests have custom datatypes and operations, e.g. allred3/red4.c (there are some more) but they have the same send and recv datatypes. What will be useful to have is a test that has different send and recv datatypes.

**Comment by akhillanger** [comment]
@Bayyapu what is the status of this PR?

**Pull request closed without merging**